styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java

```diff
@@ -102,57 +109,79 @@ public WorkflowResource(Storage storage, String schedulerServiceBaseUrl,
             rc -> patchState(arg("cid", rc), arg("wfid", rc), rc.request()))
     );
 
-    final List<Route<AsyncHandler<Response<ByteString>>>> forwardedRoutes = Arrays.asList(
-        Route.async(
-            "POST", BASE + "/<cid>",
-            rc -> createOrUpdateWorkflow(arg("cid", rc), rc)
-        ),
-        Route.async(
-            "DELETE", BASE + "/<cid>/<wfid>",
-            rc -> deleteWorkflow(arg("cid", rc), arg("wfid", rc), rc)
-        )
-    );
-
-    return cat(
-        Api.prefixRoutes(routes, V3),
-        Api.prefixRoutes(forwardedRoutes, V3)
-    );
+    return Api.prefixRoutes(routes, V3);
   }
 
-  private CompletionStage<Response<ByteString>> deleteWorkflow(String cid, String wfid,
-                                                               RequestContext rc) {
-    // TODO: handle workflow crud directly in api service instead of proxying to scheduler
-    return forwardingClient.send(rc.request().withUri(schedulerApiUrl("workflows", cid, wfid)));
+  private Response<ByteString> deleteWorkflow(String cid, String wfid) {
+    final WorkflowId workflowId = WorkflowId.create(cid, wfid);
+    final Optional<Workflow> workflow;
+    try {
+      workflow = storage.workflow(workflowId);
+    } catch (IOException e) {
+      final String message = String.format("Couldn't read workflow %s. ", workflowId);
+      LOG.warn(message, e);
+      return Response.forStatus(Status.INTERNAL_SERVER_ERROR
+          .withReasonPhrase("Error in internal storage"));
+    }
+    
+    if (!workflow.isPresent()) {
+      return Response.forStatus(Status.NOT_FOUND.withReasonPhrase("Workflow does not exist"));
+    }
+
+    try {
+      storage.delete(workflowId);
```

Comment:
Minor detail and not necessary to do in this PR but deletion should probably be atomic. I.e. do the lookup and delete in a transaction. Semantically it seems strange for two concurrent deletes to both be able to succeed due to racy existence checking and removal.

Reply:
I will work on a separate PR to solve atomic issue for both deleting and creating.